### ISSUE TITLE
Indexing a 1-D tensor fails to compile (no supported output layout)

### ISSUE THREAD

**Issue by DevAyush22** [torch-spyre/torch-spyre#4360]
### What happened?

Part of https://github.com/torch-spyre/torch-spyre/issues/3083

Indexing a one-dimensional tensor does not compile on Spyre. Any `table[index]` where `table` has rank 1 is rejected during layout selection.

The same gather with a 2-D source compiles and returns correct values, so this is specific to a rank-1 source.

| source | index | result |
| --- | --- | --- |
| `[128]` bf16 | `[24, 8]` | fails |
| `[128]` bf16 | `[1, 8]` | fails |
| `[192]` fp32 | `[192]` | fails |
| `[24, 2816]` bf16 | `[192]` | passes |

Shows up in gemma-4-26B, the per-expert scale lookup `per_expert_scale[top_k_index]`. Production models avoid it today because hf-adapters transfers that table through `dma_moe_per_expert_scale_to_spyre` instead of a plain `.to("spyre")`.

### What did you expect to happen?

`table[index]` compiles and returns the selected elements, matching a CPU reference exactly. A gather is a pure copy, so the values should come back bit-identical.

### How can we reproduce it?

```
import torch

src = torch.randn(128, dtype=torch.bfloat16)
idx = torch.randint(0, 128, (24, 8), dtype=torch.int64)
expected = src[idx]

compiled = torch.compile(lambda t, i: t[i], backend="inductor", dynamic=False)
actual = compiled(src.to("spyre"), idx.to("spyre"))   # raises Unsupported

torch.testing.assert_close(actual.to("cpu"), expected, rtol=0, atol=0)
```

### Any environmental details we need to know?

Platform: Linux, Python 3.12.13, pytest 9.1.1
Backend: `torch.compile` with inductor, device `"spyre"`

### Anything else we need to know?

Blocks the gemma-4-26B mixture-of-experts path, where three separate expert-scale lookups hit this.

### Relevant log output

```Shell
Unsupported: Spyre backend does not support: Multi-arg pointwise (buf0): no
supported output layout found with size=[24, 8],
output_ea=ElementArrangement.STANDARD, coordinates=[d0, d1].
Input EAs: [<ElementArrangement.STANDARD: 0>, <ElementArrangement.STANDARD: 0>]
```

**Comment by DevAyush22** [comment]
### Root cause

A 1-D table has only one dimension, and stickification puts it in the stick. So the index addresses a lane *inside* a stick, and the hardware only moves whole sticks — it cannot pick a lane by a value it learns at runtime. `_is_supported_layout` rejects exactly that:

```
# propagate_layouts.py
if index_sym in stick_coord.free_symbols:
    return False
```

The stick coordinate is `Mod(tmp0, 64)` where `tmp0` is the index. For a rank-1 table no `dim_order` avoids it, so every candidate is rejected and the op fails.

That guard is correct. The bug is that the pass meant to repair the layout could not.

`enforce_indirect_access_layout` exists to give an indirect-access table a workable layout, and its only tool was rotation. When the indexed dim *is* the stick, `indirect_device_pos == stick_pos` and the rotation lists it twice:

```
order = [2] + [i for i in range(3) if i != 2 and i != 2] + [2]   # [2, 0, 1, 2]
```

Rank-3 `[2, 1, 64]` becomes rank-4 `[64, 2, 1, 64]` — 8192 device slots for a 128-value tensor.

**That also produces wrong numbers silently.** Where the table is built in the graph rather than passed in, the malformed layout compiles and runs.

```
expected  [[  0,   1,   2,      63,      64,     65,    100,    127]]
actual    [[  0,   2,  -0.0013, -0.0013,  0,     1.508, 1.516,  0  ]]
```

Second half of the problem: the rejection happens in `propagate_spyre_tensor_layouts`, four passes before `enforce_indirect_access_layout` runs. So even with rotation fixed, compilation died before the repair could execute.

### Fix

Three things had to change — the wrong answers, the compile failure, and then making the result usable by anything downstream.

**1. The wrong answers.** `_build_required_stl` now re-tiles instead of rotating when the indexed coordinate is the stick: each entry gets its own stick, with the stick mapped to no host element. That is the layout a host `[N, 1]` tensor already receives, so this lands the 1-D case somewhere the backend already handles rather than somewhere new. The entry count comes from the host extent, since device dims round up to whole sticks and a 100-entry table sits in dims covering 128 slots.

A re-tile also has to actually move data. The pass preferred relabelling the producer's layout, which moves no bytes — fine for a reorder, wrong for a re-tile that changes which elements share a stick. `_is_permutation_of` now decides between the two, so rotations keep their existing free path.

**2. The compile failure.** The repair above happens late in the pipeline and the layout search gives up long before it, so by itself it never gets a chance to run.

The search rejects these tables because the index lands in the stick. That is usually the right call, but a table with only one real dimension has nowhere else to put the index, so rejecting it just means the op fails. It now lets that one case through and leaves the repair to deal with it. Tables that have another dimension to work with still get rejected here. That is fine, because the search then tries a different ordering and finds one that works.

**3. Fusing anything onto the result.** Getting `table[index]` to compile is not the whole job — `weights * table[index]` has to work too, and that needs both sides to agree about what sits in the stick.

After re-tiling they don't agree. The gather output holds one value per stick, with nothing mapped to the stick itself. An ordinary operand has a real dimension in its stick, packed 64 values deep. So the operand has to be copied into the gather's layout first.

That copy could not be planned. `compute_restickify_target_layout` only knows one move — take the stick from host dim X and put it on host dim Y — and here the target stick holds no host dim at all. There was nothing for it to match, so it returned `None`, which the planner reads as impossible:

```
elif tgt is None:
    cost = INF  # infeasible restickify
```

and reports `No mechanism to gather elements from multiple sticks into single stick`. Nothing was missing in the backend. The layout simply could not be described.

`compute_restickify_needed` now builds that target itself, collapsing any dimension the operand broadcasts over. The matmul branch already constructs an empty-stick target the same way, so this follows existing practice.

**This third gap is older than the 1-D case.** A reduction over the last dim empties the stick for the same reason a re-tiled gather does, and two tests in `test_restickify.py` were already pinning it with `pytest.raises`. For `a.amin(-1) + b`:

| tensor | host shape | device_size | stride_map |
|---|---|---|---|
| `a` | [128,128,128] | [128,2,128,64] | [128,64,16384,**1**] |
| `a.amin(-1)` | [128,128] | [128,1,128,64] | [1,-1,128,**-1**] |
| `b` | [128,128] | [2,128,64] | [64,128,**1**] |
| `b` restickified | [128,128] | [128,1,128,64] | [1,-1,128,**-1**] |

The reduction consumes the host dim that was living in `a`'s stick, leaving nothing to put there. `b` is dense. They disagree, so `b` has to be restickified — and that was exactly the target the compiler could not express. Both tests now compile and give correct results, so they check the result instead of the failure. Removing the fix brings the original error back.

### Why this never showed up in hf-adapter

hf-adapter does not let the table reach the device in that shape. For this tensor it calls `dma_moe_per_expert_scale_to_spyre` from `model_utils.py`, which replicates each scale 64 times on the host so every entry fills a stick. Its docstring is explicit about why:

> Each scale fills one stick. Widening on the host avoids an unsupported in-graph rank expansion.

It works, and it was a reasonable thing to do given the compiler could not. But it only helps tensors you own before compiling, it needs the model author to know about sticks, and — relevant here — by steering every real workload off the broken path it kept the silent wrong-answer defect alive and unnoticed.

Once this lands, `dma_moe_per_expert_scale_to_spyre`  should no longer be needed for correctness in hf-adapter.

### Known limits

- **Footprint.** One stick per entry costs 128 bytes regardless of dtype, so the table grows 64× at bfloat16. Practical ceiling is around 2M entries. Past the span limit it refuses with a reason instead of failing deep in the backend:

  ```
  Unsupported: cannot re-tile value tensor: one entry per stick would take
  512.00 MB for 4194304 entries, over the 256.00 MB per-core limit. Give the
  tensor a trailing dimension so its entries already sit on separate sticks.
  ```

- **float32 still fails**, but no longer on layout. It now gets past layout selection and reaches `ReStickifyOpHBM on DataFormats.IEEE_FP32`, which is the gap already tracked in #2544 and #2560. Pre-existing and unrelated to this change.
